### PR TITLE
Show unit price in shopfront

### LINKED ISSUE #6497: Show unit price in Shopfront
## Description

**- As a:** user
**- On page:** <enterprise_name>/shop#/shop
**- I want to be able to do:** See the unit price of each of the variants of a product, including the bulk buy modal.

## Acceptance Criteria & Tests

The unit price should include any fees listed in the price breakdown. That is, all fees associated with a variant. See [app/assets/javascripts/templates/price_breakdown.html.haml](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/assets/javascripts/templates/price_breakdown.html.haml) and [Api::VariantSerializer#fees](https://github.com/openfoodfoundation/openfoodnetwork/blob/f2c37ef64d05bd797ea54fc72acd619ed6879d2a/app/serializers/api/variant_serializer.rb#L10-L13) for reference.

Check the design specs linked in https://github.com/openfoodfoundation/openfoodnetwork/issues/6449#issue-756226197. Watch out mobile!

1. Go to an enterprise's shopfront.
2. When the product is bought in bulk, click on "Add".

The unit price should correctly match the price displayed for the variant in both the variant and the bulk modal.

### PR DESCRIPTION
**Pull request by jibees**
#### What? Why?
Related to #6497
Display a new (fake) information on a product variant: unit price.

_EDITED:_ Enabled for both staging and development environment. 

#### What should we test?
You must see into the product list of a shop a new information with its tooltip. 
NB: this information is still fake.

Capture d’écran 2021-02-16 à 15 25 32

In production, nothing should happen.

#### Release notes
Display unit price information for a product in the shopfront.

Changelog Category: User facing changes

### CODE REVIEW

File: app/views/shop/products/_form.html.haml
Comment (sauloperez) [inline app/views/shop/products/_form.html.haml:18]:
```suggestion
                  - else
```

File: app/serializers/api/variant_serializer.rb
Comment (sauloperez) [inline app/serializers/api/variant_serializer.rb:49]: I don't mean to change it because this is just temporal but I was curious and went check Ruby docs and the `Random` class has a way to do this using `#rand(max)` :point_right: https://ruby-doc.org/core-2.4.0/Random.html#method-c-rand :ok_hand:

File: app/assets/stylesheets/darkswarm/question-mark-icon.scss
Comment (sauloperez) [inline app/assets/stylesheets/darkswarm/question-mark-icon.scss:50]: what a mess... I feel your pain :joy:

File: app/assets/stylesheets/darkswarm/question-mark-icon.scss
Comment (sauloperez) [inline app/assets/stylesheets/darkswarm/question-mark-icon.scss:43]: it'd be great to standardize this radius, don't you think? I don't know how feasible this is though but with a quick git grep I see we use 3px quite a lot and we did the first attempt in 

https://github.com/openfoodfoundation/openfoodnetwork/blob/5fd1bc3604f50884c749bf3d2a202697cad343aa/app/assets/stylesheets/admin/globals/variables.scss#L147

although for /admin only

File: app/views/shop/products/_form.html.haml
Comment (jibees) [reply app/views/shop/products/_form.html.haml]: Resolved: https://github.com/openfoodfoundation/openfoodnetwork/pull/6851/commits/e2be6e0c4f2f41c50bc53428b731a6695ee29884#diff-23c17670591779ec3cc7524409bc47fa9d184169b91c9bc40989cdef7487bb36R18

**Review by sauloperez — APPROVED** [review]
Amazing work @jibees ! :clap:

**Comment by sauloperez** [comment]
`OFN_FEATURE_UNIT_PRICE: "true"` shouldn't be necessary @jibees . That's the earlier ENV-based feature toggles we had which are still in use. That needs to made clear in the code, that's true.

File: app/assets/stylesheets/darkswarm/question-mark-icon.scss
Comment (jibees) [reply app/assets/stylesheets/darkswarm/question-mark-icon.scss:43]: I'm wondering the possibility, the avantages, the drawbacks to put a utility-first css framework (such as tailwind...) on top of the front side of the website to answer this kind of issue (using class, see : https://tailwindcss.com/docs/border-radius)

Comment (jibees) [reply app/assets/stylesheets/darkswarm/question-mark-icon.scss:43]: I know we're not running last angular ... but: https://twitter.com/angular/status/1359736376581840896

Comment (sauloperez) [reply app/assets/stylesheets/darkswarm/question-mark-icon.scss:43]: I dream about it. Tailwind made ❤️ CSS again but it feels daunting to migrate to it at this point

**Comment by filipefurtad0** [comment]
Hey @jibees ,

Cool, thanks for pointing out the need to enable it through the feature toggle.

I had a look at a given shopfront before:

image

And after staging and enabling this feature:

image

The tooltip is also displayed correctly:

image

If I understand correctly, this is all which needs to be tested right now (the pics shown above). In that case this would be ready to go. If there is something else which should be tested, please let me know :+1:

**Comment by jibees** [comment]
I don't know if @Erioldoesdesign wants to check the screenshot if it suits.

**Comment by filipefurtad0** [comment]
Thanks @jibees ,
Moved it back to Test-ready, so it doesn't get merged before feedback. Updated also with pics from iPhone and Android.

**Comment by jibees** [comment]
Just realized (thanks to the screenshot that the `border-radius` is invalid, and so it's not compliant with @Erioldoesdesign specification. Moving to In Dev, will do the modification.

File: app/assets/stylesheets/darkswarm/question-mark-icon.scss
Comment (jibees) [reply app/assets/stylesheets/darkswarm/question-mark-icon.scss]: I used `$radius-small` variable. 
https://github.com/openfoodfoundation/openfoodnetwork/pull/6851/files#diff-704a4699e3348b7fe0b6794127d7ac67638ca1b5394f5938a9b5cc2f2eff0035R44

**Comment by jibees** [comment]
I've made to small modifications in the CSS files, which should not impact the code review, but impact the final design with small border radius as you can see: 
Capture d’écran 2021-02-12 à 09 38 07
Moving to Test Ready

**Comment by filipefurtad0** [comment]
Thanks @jibees for addressing this so quickly! 

Just to make sure I'm looking at the right design specifications, do you mean this link:
https://www.figma.com/file/EPKyXwl5VDQaw8IBHalCEZ/Unit-Prices?node-id=0%3A1

Thanks for spotting that - I see the rounded corners now as well :+1: 

Ok, after these latest commits:

**Desktop**
image

**iPhone 11 - Safari**

image


Firefox on Samsung Galaxy S20

Please note that the text is slightly un-centered, when using the on top view, and on narrower screens (I don't think this was the case yesterday, or maybe I simply did not notice it?):

image

But maybe ok to merge like this, especially since this feature will not be visible to all users just yet - only after toggling it. What do you think JB?

**Comment by sauloperez** [comment]
Something I'm noticing on those screenshots is that there's not enough margin between the text and its containing blue box. IMO it needs to breathe a bit more. We can check that offline with @Erioldoesdesign and iterate on it in new PR. I'm starting to love implementing features with feature toggles :heart_eyes:

File: config/initializers/feature_toggles.rb
Comment (jibees) [inline config/initializers/feature_toggles.rb:14]: @sauloperez 
I just had a look to https://github.com/openfoodfoundation/ofn-install/pull/704 and if I understood correctly, each user of FR instance is now considered as a `beta_tester` which is quite embarrassing for the unit_price feature. Actually, as I coded, I'm waiting for a list of emails.
I'm wondering if we should consider to use a `env` variable per feature? Such as `UNIT_PRICE_BETA_TESTERS` and have a unique management system for each case?

Comment (sauloperez) [reply config/initializers/feature_toggles.rb:14]: but notice that due to that the condition `beta_testers.include?(user.email)` will always evaluate to false. Indeed, it's something we need to solve differently but it won't reveal the feature.

Comment (jibees) [reply config/initializers/feature_toggles.rb:14]: - I've created an issue to solve the _list of emails_: #6870
 - For the second point, I guess I should have a look at `ENV['RAILS_ENV']` or something like that? 

Moving to `In Dev`.... ;-)

File: config/initializers/feature_toggles.rb
Comment (jibees) [inline config/initializers/feature_toggles.rb:14]: I'm asking for your review @sauloperez ! Thanks. 🙏

File: config/initializers/feature_toggles.rb
Comment (jibees) [reply config/initializers/feature_toggles.rb]: Modification is done: https://github.com/openfoodfoundation/openfoodnetwork/pull/6851/commits/406c18051146169b6de567bc8b8cf84a5a1ec0e5#diff-0f07cf1c313099baff76848102decb21d55a569b13e0aedb05f6b855f684cf35R14

I hope this is what was expected.

Moving back to `Code Review` waiting for @sauloperez approval.

**Comment by jibees** [comment]
I had to do some other _under the hood_ modification, so I took advantage of it to do some appearance modification, as asked: 
Capture d’écran 2021-02-12 à 17 24 31

**Comment by andrewpbrett** [comment]
This looks good to me - but I don't think it should close the issue until the real unit prices are appearing, right?

**Comment by Erioldoesdesign** [comment]
This looks good re. rounded corners and breathing space between blue-box edge and text:
image

I also notice that the '?' in the circle appears off-centre. Is that because of the icon overall size?

**Comment by sauloperez** [comment]
> but I don't think it should close the issue until the real unit prices are appearing, right?

yes, good catch. #6497 still needs its backend. I updated the description accordingly.

**Comment by RachL** [comment]
Shouldn't we have 2 issues? One for the front-end that can be closed and a second one for backend?

**Comment by jibees** [comment]
> I also notice that the '?' in the circle appears off-centre. Is that because of the icon overall size?

Yes, it's mainly because it's purely CSS. I think we need here a SVG icon. Should I open another issue to resolve it?


> > but I don't think it should close the issue until the real unit prices are appearing, right?
> 
> yes, good catch. #6497 still needs its backend. I updated the description accordingly.

Thanks! I think it's more clear, and it allows us to merge this issue and continue the implementation on other pages.

File: config/initializers/feature_toggles.rb
Comment (sauloperez) [reply config/initializers/feature_toggles.rb:14]: yes, that also feels enough at this stage to me. Btw, FYI you can also use Ruby's [word array operator](https://stackoverflow.com/a/1274703/2838228) or Rails' [StringInquierer](https://stackoverflow.com/a/2715076/2838228). Just information, I don't mean to change it.

**Comment by jibees** [comment]
> Shouldn't we have 2 issues? One for the front-end that can be closed and a second one for backend?

I agree. I will create the backend issue.

**Review by sauloperez — APPROVED** [review]
I'd be happy to merge this, show it to @Erioldoesdesign in staging, and have her raise any design.

**Comment by jibees** [comment]
@RachL @sauloperez @andrewpbrett 
Issue for backend part created: #6877

**Comment by Erioldoesdesign** [comment]
@jibees I can make/provide an SVG if someone makes the issue and assigns me?

Also, @sauloperez let me know what staging site to check on and I can 'design QA' 😛

**Comment by jibees** [comment]
> @jibees I can make/provide an SVG if someone makes the issue and assigns me?

@Erioldoesdesign consider it done ;) #6878

**Comment by jaycmb** [comment]
Just noticed when looking at the tooltip, we still have the initial placeholder text in there.
It should better be phrased:
"This is the unit price of this **product**
The unit price allows you to compare the price of products **independent of packaging sizes & weights**"

It´s not visible to the shopper yet, but let´s rather fix it while on it anyways :)

**Comment by sauloperez** [comment]
This got checked by @jaycmb in staging so we can move on in its implementation :muscle:

**Comment by jaycmb** [comment]
Okay, quick feedback from staging: 
**Tooltip**
Design visually looks good (like margins & line breaks with new text), but the **behaviour** of the tool tip is a bit odd:
- It´s not possible to close it, expect for user clicks on the tooltip again -> that´s quite annoying on mobile especially, since it´s gonna be hard to access. 

Not sure if the following was different before and approved by @Erioldoesdesign , but I would have expected:
- Tooltip should appear only on hover (Desktop) and Tap on mobile, but otherwise click on desktop also fine
- Tooltip disappears when user taps anywhere else (on mobile) or stops hovering (on desktop)

**Spacing Mobile:**

The Unit Price div needs more space to the _add to cart_ and especially the _- + element_.


image

**Comment by RachL** [comment]
So we now have a process for product review after development?

IMO if the issue is well described, there is no need for product or design review, the tester has all the info to make it go forward... If something was forgotten / unclear,like the feedback I see here, it belongs to a new issue. Otherwise I'm afraid we end up with a scope creep that requires even to do the code review step again...

Happy to be proven wrong, but I just wanted to flag that this step does not exist in our processes today @sauloperez  @jaycmb @Erioldoesdesign

**Comment by jaycmb** [comment]
No, that´s no new process :) 
Because of the feature toggle, it was proposed to merge the PR even if it was still waiting for design feedback, because it was blocking other work on the epic and fine-tuning can be done later.
Instead of doing so, I wanted to quickly check if there are any obvious things - I think @sauloperez commented/concluded to early that it has been checked & approved - than should be addressed now.
I think we have to get clear on the way how the feature toggle is used related to testing & final approval.

**Comment by filipefurtad0** [comment]
I see this PR is now merged, so I'll disconnect the main issue #6497 and move tit to In dev.

**Comment by Erioldoesdesign** [comment]
@jaycmb The new unit prices tooltip should function exactly like the pie tooltip. Which isn't great but doesn't scope creep complexity re. development.

The pie you can click 'out' of the tooltip and it closes it or you click on the icon (which actually changes to an 'X' in the circle but it's not very visible.

The unit price tooltip should be on click to mimic existing functionality re. pie icon. We shouldn't have one that hovers and one that is click.

So:

> But also fine with:
> tooltip appears on click on question mark (both desktop & mobile) and closes when user clicks anywhere outside of tooltip area (not just on question mark icon)

That is what I agree with and intended re. design.

Agreed the spacing you have is not the same as the design on mobile which looks like:

Screenshot 2021-02-22 at 15 25 48

(the  - / + button is an old version from Yuko's designs so ignore that part please)

Re. @RachL and a new process...I assumed this stage was like the 'design sweep' that Yuko and Danni used to do and not new. We've spoken briefly in some meetings about Design QA (or Design sweep) and how design can work with testing but unsure who and how to progress these conversations.

**Comment by RachL** [comment]
@Erioldoesdesign the testing team is meeting every 2 weeks, maybe we can add the topic to our meeting on March 9th and you can join us to have a chat on it?
I didn't know Yuko and Danni did this. I only thought that Danni (and sometimes Jen) did all the testing on the mobile feature. The testing team therefore was not involved except at the end when Danni did not have time for mobile anymore.  I would like to understand why we feel it is needed and I think it would be great to run a retro to understand how the team feels about it.

**Comment by Erioldoesdesign** [comment]
@RachL sure thing! I'll join the testing team meeting on the 9th 😄

**Pull request merged**
